File: app/src/main/java/com/originb/inkwisenote/modules/functionalUtils/Either.java
package com.originb.inkwisenote.modules.functionalUtils;

public class Either<Err, Res> {
    public Err error;
    public Res result;

    public static <Res> Either result( Res value) {
        return new Either(null, value);
    }

    public static <Err> Either error( Err error) {
        return new Either(error, null);
    }

    private Either(Err err, Res value) {
        this.error = err;
        this.result = value;
    }
}

File: app/src/main/java/com/originb/inkwisenote2/functionalUtils/Either.kt
package com.originb.inkwisenote2.functionalUtils

class Either<Err, Res> private constructor(var error: Err, var result: Res) {
    companion object {
        fun <Res> result(value: Res): Either<*, *> {
            return Either<Any?, Any?>(null, value)
        }

        fun <Err> error(error: Err): Either<*, *> {
            return Either<Any?, Any?>(error, null)
        }
    }
}
